File: mariadb-api/src/Controllers/ReleasesController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Services\ReleaseService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/** Ported from Controllers/ReleasesController.cs. */
final class ReleasesController extends BaseController
{
    private function service(): ReleaseService
    {
        return new ReleaseService(Database::connection());
    }

    public function create(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->create($args['projectId'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function update(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->update($args['projectId'], $args['id'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function delete(Request $request, Response $response, array $args): Response
    {
        return $this->service()->delete($args['projectId'], $args['id']) ? $this->noContent($response) : $this->notFound($response);
    }

    // ReleaseNotes is the Release Notes Packager's own admin-only field — gated by
    // ProjectAdminMiddleware on its own route sub-group in routes.php (which already means
    // "Project Admin OR Org Admin"), never writable via the generic update() action above.
    public function updateNotes(Request $request, Response $response, array $args): Response
    {
        $body = $this->body($request);
        $result = $this->service()->updateNotes($args['projectId'], $args['id'], $body['releaseNotes'] ?? null);
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }
}

File: api/Enkl.Api/Controllers/ReleasesController.cs
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Enkl.Api.Controllers;

[ApiController]
[Authorize(Policy = "ProjectMember")]
[Route("api/projects/{projectId:guid}/releases")]
public class ReleasesController : ControllerBase
{
    private readonly ReleaseService _releases;

    public ReleasesController(ReleaseService releases)
    {
        _releases = releases;
    }

    [HttpPost]
    public async Task<IActionResult> Create(Guid projectId, CreateReleaseRequest request)
    {
        var result = await _releases.CreateAsync(projectId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPut("{releaseId:guid}")]
    public async Task<IActionResult> Update(Guid projectId, Guid releaseId, UpdateReleaseRequest request)
    {
        var result = await _releases.UpdateAsync(projectId, releaseId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{releaseId:guid}")]
    public async Task<IActionResult> Delete(Guid projectId, Guid releaseId)
    {
        return await _releases.DeleteAsync(projectId, releaseId) ? NoContent() : NotFound();
    }

    // ReleaseNotes is the Release Notes Packager's own admin-only field — this action overrides the
    // class-level ProjectMember policy with ProjectAdmin (which already means "Project Admin OR Org
    // Admin", see ProjectAdminAuthorizationHandler.cs), rather than letting any project member write
    // it via the generic Update action above.
    [HttpPut("{releaseId:guid}/notes")]
    [Authorize(Policy = "ProjectAdmin")]
    public async Task<IActionResult> UpdateNotes(Guid projectId, Guid releaseId, UpdateReleaseNotesRequest request)
    {
        var result = await _releases.UpdateNotesAsync(projectId, releaseId, request);
        return result is null ? NotFound() : Ok(result);
    }
}
